pkgs/development/compilers/crystal/default.nix

```diff
@@ -1,93 +1,142 @@
+        "x86_64-darwin" = "0hzc65ccajr0yhmvi5vbdgbzbp1gbjy56da24ds3zwwkam1ddk0k";
+      }."${stdenv.system}";
+    };
+
+    buildCommand = ''
+      mkdir -p $out
+      tar --strip-components=1 -C $out -xf ${src}
+    '';
   };
 
-  unpackPhase = ''
-    mkdir ${prebuiltName}
-    tar --strip-components=1 -C ${prebuiltName} -xf ${prebuiltSrc}
-    tar xf ${src}
-  '';
-
-  # crystal on Darwin needs libiconv to build
-  libs = [
-    boehmgc libatomic_ops pcre libevent
-  ] ++ stdenv.lib.optionals stdenv.isDarwin [
-    libiconv
-  ];
-
-  nativeBuildInputs = [ which makeWrapper ];
-
-  buildInputs = libs ++ [ llvm ];
-
-  libPath = stdenv.lib.makeLibraryPath libs;
-
-  sourceRoot = "${name}";
-
-  preBuild = ''
-    patchShebangs bin/crystal
-    patchShebangs ../${prebuiltName}/bin/crystal
-    export PATH="$(pwd)/../${prebuiltName}/bin:$PATH"
-  '';
-
-  makeFlags = [ "CRYSTAL_CONFIG_VERSION=${version}"
-                "FLAGS=--no-debug"
-                "release=1"
-                "all" "docs"
-              ];
-
-  installPhase = ''
-    install -Dm755 .build/crystal $out/bin/crystal
-    wrapProgram $out/bin/crystal \
-        --suffix PATH : ${clang}/bin \
-        --suffix CRYSTAL_PATH : lib:$out/lib/crystal \
-        --suffix LIBRARY_PATH : $libPath
-    install -dm755 $out/lib/crystal
-    cp -r src/* $out/lib/crystal/
-
-    install -dm755 $out/share/doc/crystal/api
-    cp -r docs/* $out/share/doc/crystal/api/
-    cp -r samples $out/share/doc/crystal/
-
-    install -Dm644 etc/completion.bash $out/share/bash-completion/completions/crystal
-    install -Dm644 etc/completion.zsh $out/share/zsh/site-functions/_crystal
-
-    install -Dm644 man/crystal.1 $out/share/man/man1/crystal.1
-
-    install -Dm644 LICENSE $out/share/licenses/crystal/LICENSE
-  '';
-
-  dontStrip = true;
+  generic = { version, sha256, doCheck ? true }:
+  stdenv.mkDerivation rec {
+    inherit doCheck;
+    name = "crystal-${version}";
+
+    src = fetchFromGitHub {
+      owner  = "crystal-lang";
+      repo   = "crystal";
+      rev    = version;
+      inherit sha256;
+    };
+
+    # the first bit can go when https://github.com/crystal-lang/crystal/pull/6788 is merged
+    postPatch = ''
+      substituteInPlace src/compiler/crystal/config.cr \
+        --replace '{{ `date "+%Y-%m-%d"`.stringify.chomp }}' '"${releaseDate}"'
+      ln -s spec/compiler spec/std
+      substituteInPlace spec/std/process_spec.cr \
+        --replace /bin/ /run/current-system/sw/bin
+    '';
+
+    buildInputs = [
+      boehmgc libatomic_ops pcre libevent
+      llvm
+    ] ++ stdenv.lib.optionals stdenv.isDarwin [
+      libiconv
+    ];
+
+    nativeBuildInputs = [ binary makeWrapper which ];
+
+
+    makeFlags = [
+      "CRYSTAL_CONFIG_BUILD_DATE=${releaseDate}"
+      "CRYSTAL_CONFIG_VERSION=${version}"
+    ];
+
+    buildFlags = [
+      "all" "docs"
+    ];
+
+    FLAGS = [
+      "--release"
+      "--single-module" # needed for deterministic builds
```

Comment:
@peterhoeg Do you recall why this `--single-module` is/was necessary? This flag is still used and I find it a bit odd as I don't see why this would make a difference.
It's quite inefficient to use `--single-module` because LLVM codegen cannot be parallelized.

Reply:
I unfortunately do not remember, sorry.

Just FYI, I'm mentoring a gsoc project to revamp the crystal support in nix so there are changes coming over the summer.

Reply:
Oh that's nice to hear. I'm happy to help if you need any upstream support or expertise.